src/pkcs11/pkcs11-spy.c

```diff
@@ -400,6 +400,116 @@ spy_attribute_list_out(const char *name, CK_ATTRIBUTE_PTR pTemplate,
 	print_attribute_list(spy_output, pTemplate, ulCount);
 }
 
+static void
+spy_dump_mechanism_in(const char *name, CK_MECHANISM_PTR pMechanism)
+{
+	if (!pMechanism) {
+		fprintf(spy_output, "[in] %s=(null)\n", name);
+		return;
+	}
+
+	fprintf(spy_output, "[in] %s->type=%s\n", name, lookup_enum(MEC_T, pMechanism->mechanism));
+	switch (pMechanism->mechanism) {
+	case CKM_AES_GCM:
+		if (pMechanism->pParameter != NULL) {
+			CK_GCM_PARAMS *param =
+				(CK_GCM_PARAMS *) pMechanism->pParameter;
+			spy_dump_string_in("pIv[ulIvLen]",
+				param->pIv, param->ulIvLen);
+			spy_dump_ulong_in("ulIvBits", param->ulIvBits);
+			spy_dump_string_in("pAAD[ulAADLen]",
+				param->pAAD, param->ulAADLen);
+			fprintf(spy_output, "[in] %s->pParameter->ulTagBits=%lu\n", name, param->ulTagBits);
+		} else {
+			fprintf(spy_output, "[in] %s->pParameter = NULL\n", name);
+			break;
+		}
+		break;
+	case CKM_ECDH1_DERIVE:
+	case CKM_ECDH1_COFACTOR_DERIVE:
+		if (pMechanism->pParameter != NULL) {
+			CK_ECDH1_DERIVE_PARAMS *param =
+				(CK_ECDH1_DERIVE_PARAMS *) pMechanism->pParameter;
+			fprintf(spy_output, "[in] %s->pParameter = {\n\tkdf=%s\n", name,
```

Comment:
In here we have `= {` form. And in above we do not have this.

Is there way that you prefer -- I can sync them too.

Reply:
That would be great. I think the braces are nice and avoid repeating the `pMechanism->pParameter`, but indentation with tabs might not always look like you would expect especially with mutli-line parameters so I would prefer to use the syntax from AES_GCM and always prefix the nested values with `pMechanism->pParameter`.